src/credentials/GroupDataProvider.h

```diff
@@ -0,0 +1,236 @@
+    // multiple GroupMapping having the same `group` value, but each with different
+    // `endpoint` value.
+    struct GroupMapping
+    {
+        // The endpoint to which a GroupId is mapped.
+        EndpointId endpoint;
+        // The GroupId, which, when received in a message will map the the `endpoint`.
+        GroupId group;
+
+        bool operator==(const GroupMapping & other) { return this->endpoint == other.endpoint && this->group == other.group; }
+
+        GroupMapping & operator=(const GroupMapping & other)
+        {
+            if (this == &other)
+                return *this; // Guard self assignment
+            this->endpoint = other.endpoint;
+            this->group    = other.group;
+            return *this;
+        }
+    };
+
+    // A group state maps the group key set to use for encryption/decryption for a given group ID.
+    struct GroupState
+    {
+        // Fabric Index associated with the group state entry's fabric scoping
+        chip::FabricIndex fabric_index;
+        // Identifies the group within the scope of the given fabric
+        chip::GroupId group;
+        // References the set of group keys that generate operationa group keys for use with the given group
+        uint16_t key_set_index;
+    };
+
+    // A operational group key set, usable by many GroupState mappings
+    struct KeySet
+    {
+        enum class SecurityPolicy : uint8_t
+        {
+            kStandard   = 0,
+            kLowLatency = 1
+        };
+
+        // The actual keys for the group key set
+        EpochKey epoch_keys[3];
+        // Logical index provided by the Administrator that configured the entry
+        uint16_t key_set_index;
+        // Security policy to use for groups that use this keyset
+        SecurityPolicy policy;
+        // Number of keys present
+        uint8_t num_keys_used;
+    };
+
+    // Iterator for group mappings under a given endpoint. Associated with
+    // Groups cluster logic.
+    class GroupMappingIterator
+    {
+    public:
+        virtual ~GroupMappingIterator() = default;
+        // Returns the number of entries in total that will be iterated.
+        virtual uint16_t Count() = 0;
+        // Returns true if a groupID is found in the iteration.
+        virtual bool Next(GroupId & outGroup) = 0;
+        // Release the memory allocated by this iterator, if any. Must be called before
+        // losing scope of a `GroupMappingIterator *`
+        virtual void Release() = 0;
+
+    protected:
+        GroupMappingIterator() = default;
+    };
+
+    // Iterator for group state information mapping a Group ID to a Group Key Set index,
+    // such as reflected by the Groups attribute of the Group Key Management cluster.
+    class GroupStateIterator
+    {
+    public:
+        virtual ~GroupStateIterator() = default;
+        // Returns the number of entries in total that will be iterated.
+        virtual uint16_t Count() = 0;
```

Comment:
Likewise suggest size_t for this.  Once we consider multi-fabric, I don't think this is actually bound to UINT16_MAX.